dockstore-webservice/src/main/java/io/dockstore/webservice/resources/WorkflowResource.java

```diff
@@ -2321,50 +2323,95 @@ public List<Workflow> updateDois(@Parameter(hidden = true) @Auth User user,
+                try {
+                    if (updateWorkflowWithDois(workflow, gitHubRepoDoi)) {
                         updatedWorkflows.add(workflow);
                     }
-                    // Add version DOI(s) to new or existing Concept DOI.
-                    gitHubRepoDoi.tagAndDoi().stream().forEach(tagAndDoi -> {
-                        final WorkflowVersion workflowVersion = workflowVersionDAO.getWorkflowVersionByWorkflowIdAndVersionName(
-                                workflow.getId(), tagAndDoi.gitHubTag());
-                        if (workflowVersion != null) {
-                            if (workflowVersion.getDois().get(Doi.DoiInitiator.GITHUB) == null) {
-                                final Doi versionDoi = ZenodoHelper.getDoiFromDatabase(Doi.DoiType.VERSION, Doi.DoiInitiator.GITHUB,
-                                        tagAndDoi.doi());
-                                workflowVersion.getDois().put(Doi.DoiInitiator.GITHUB, versionDoi);
-                                updatedWorkflows.add(workflow);
-                            }
-                        }
-                    });
-                    if (noDoiToStart) {
-                        // The default DOI selection is USER. If there was no DOI to begin with, set it to GITHUB so it will show up in the UI.
-                        workflow.setDoiSelection(Doi.DoiInitiator.GITHUB);
-                    }
+                } catch (Exception e) {
+                    LOG.error("Error updating workflow %s with DOIs".formatted(workflow.getWorkflowPath()), e);
                 }
             });
         });
         return updatedWorkflows.stream().toList();
     }
 
+    /**
+     * Updates a workflow with DOIs discovered from Zenodo for a GitHub repository.
+     *
+     * <ul>
+     *     <li>If the workflow doesn't have a concept DOI for the GitHub initiator adds it, then updates the versions with the version DOIs</li>
+     *     <li>If the workflow's existing concept DOI for the GitHub initiator matches the discovered concept DOI, then updates the version DOIs</li>
+     *     <li>If the workflow's exising concept DOI for the the GitHub initiator DOES NOT match the discovered concept DOI, does nothing. Because
+     *     we do not support multiple concept DOIs from the same initiator, and it doesn't make sense to have version DOIs that don't
+     *     match the concept DOI.</li>
+     * </ul>
+     *
+     * @param workflow
+     * @param gitHubRepoDoi
+     * @return true if the workflow was updated, false otherwise
+     */
+    private boolean updateWorkflowWithDois(Workflow workflow, GitHubRepoDois gitHubRepoDoi) {
+        boolean updatedWorkflow = false;
+        final String conceptDoi = gitHubRepoDoi.conceptDoi();
+        final Doi existingGitHubDoi = workflow.getConceptDois().get(DoiInitiator.GITHUB);
+        if (existingGitHubDoi != null && !conceptDoi.equals(existingGitHubDoi.getName())) {
```

Comment:
Not blocking but trying to understand: does the second part of the conditional check the DOI itself? i.e. if they update the DOI on GitHub would we catch that update?

Reply:
If the workflow already has concept DOI A associated with it, and we now detect a new concept DOI B associated with it, this code ignores DOI B.

The Dockstore model currently only supports 1 DOI per initiator, in other words we can only save 1 DOI from the Zenodo-GitHub integration.

This code is for the hopefully edge/never case when a second concept DOI gets issued via the Zenodo-GitHub integration. Our choices are:

1. Ignore the second DOI, what I do now.
2. Replace the first DOI with the second DOI. You could argue it both ways, but I feel it makes more sense to keep the first one; it would be confusing if it switches out from under you.
3. Update the model to support multiple Zenodo-GitHub DOIs.

Hopefully this rarely/never occurs, but we can see how often the message gets logged.